### PR TITLE
Add metrics spec supplementary guidelines

### PR DESCRIPTION
**Pull request by reyang**
Related to #1873 and #1891.

If you find it hard to look at the raw markdown file, here is a rendered version https://github.com/reyang/opentelemetry-specification/blob/reyang/guideline/specification/metrics/supplementary-guidelines.md.

### CODE REVIEW

File: specification/metrics/supplementary-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  * ProcessId = `1001`, ThreadId = `2`, PageFaults = `42`
+* During the time range (T<sub>3</sub>, T<sub>4</sub>]:
+  * ProcessId = `1001`, ThreadId = `1`, PageFaults = `60`
+  * ProcessId = `1001`, ThreadId = `2`, PageFaults = `47`
+* During the time range (T<sub>4</sub>, T<sub>5</sub>]:
+  * thread 1 died, thread 3 started
+  * ProcessId = `1001`, ThreadId = `2`, PageFaults = `53`
+  * ProcessId = `1001`, ThreadId = `3`, PageFaults = `5`
+
+If we export the metrics using **Cumulative Temporality**, it is quite
```
Comment (utpilla) [inline specification/metrics/supplementary-guidelines.md:152]: Should this be Delta Temporality and the next paragraph be Cumulative?

File: specification/metrics/supplementary-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+impossible** to reliably deduce the Delta temporality from Cumulative
+temporality. For example:
+
+* If the maximum value is 10 during (T<sub>0</sub>, T<sub>2</sub>] and the
+  maximum value is 20 during (T<sub>0</sub>, T<sub>3</sub>], we know that the
+  maximum value during (T<sub>2</sub>, T<sub>3</sub>] must be 20.
+* If the maximum value is 20 during (T<sub>0</sub>, T<sub>2</sub>] and the
+  maximum value is also 20 during (T<sub>0</sub>, T<sub>3</sub>], we wouldn't
+  know what the maximum value is during (T<sub>2</sub>, T<sub>3</sub>], unless
+  we know that there is no value (count = 0).
```
Comment (utpilla) [inline specification/metrics/supplementary-guidelines.md:174]: Maybe reword this to be more specific?

"unless we know that there is no value during (T<sub>2</sub>, T<sub>3</sub>] (count = 0) in which case the maximum value during (T<sub>2</sub>, T<sub>3</sub>] will be 0."

File: specification/metrics/supplementary-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    `5 (ms)`
+* (T<sub>0</sub>, T<sub>5</sub>]
+  * dimensions: {verb = `GET`, status = `200`}, count: `6`, min: `30 (ms)`, max:
+    `100 (ms)`
+  * dimensions: {verb = `GET`, status = `500`}, count: `3`, min: `1 (ms)`, max:
+    `5 (ms)`
+
+If we choose **Cumulative Temporality**, the SDK **has to track what has
+happened prior to the latest collection/export cycle**, in this worst case, the
+SDK **will remember what has happened since the ever beginning of the process**.
```
Comment (utpilla) [inline specification/metrics/supplementary-guidelines.md:101]: nit: SDK **will have to remember what has happened ever since the beginning of the process**.

File: specification/metrics/supplementary-guidelines.md (hunk 1 above)
Comment (reyang) [reply specification/metrics/supplementary-guidelines.md:152]: Nope, this is Cumulative (we're doing Cumulative->Cumulative here).

The next paragraph is about Cumulative->Delta conversion.

File: specification/metrics/supplementary-guidelines.md (hunk 2 above)
Comment (reyang) [reply specification/metrics/supplementary-guidelines.md:174]: Probably no? I think if count=0, the maximum value technically should not be exist, or it can be NaN.

File: specification/metrics/supplementary-guidelines.md (hunk 3 above)
Comment (reyang) [reply specification/metrics/supplementary-guidelines.md:101]: Fixed, thanks!

File: specification/metrics/supplementary-guidelines.md (hunk 2 above)
Comment (utpilla) [reply specification/metrics/supplementary-guidelines.md:210]: If the maximum value shouldn't technically exist, then maybe we should get rid of the "unless we know ..." part.

**Review by jsuereth — COMMENTED** [review]
This generally looks very good.  I need to take a second pass on grammar/phrasing (That's my own worst enemy), but the content looks like a great addition/treatment of the issue for implementers.

One thing I'd like to call out though is a suggestion around Memory Management and Allocations.  While it's impossible to remove all allocations on the hot path of synchronous metric collection, particularly if using Baggage, SDK authors should be encouraged to reduce memory contention *in the hot path*, and pay most overhead at startup.   This aligns with OTel's vision of "least disruption to runtime", which shows up in recommendations around error-reporting during operation and likely should also apply to things like memory usage.

I like the idea of limiting memory as suggested here, just want to also reinforce this "allocate ahead of time and reuse" notion, which we actually expose a bit more directly in Java (and I need to annotate).


Specifically, one scenario to call out (more important for Logs/Traces than metrics) is how a system behaves when memory can no longer be allocated.   Do we still get observability or is it gone?

**Comment by reyang** [comment]
> I like the idea of limiting memory as suggested here, just want to also reinforce this "allocate ahead of time and reuse" notion, which we actually expose a bit more directly in Java (and I need to annotate).

I've updated the PR to cover the pre-allocation for garbage collected languages. PTAL.
 
> Specifically, one scenario to call out (more important for Logs/Traces than metrics) is how a system behaves when memory can no longer be allocated. Do we still get observability or is it gone?

I've updated the PR trying to scratch the surface. If we need to dig deeper, a separate PR might be better?

**Review by jsuereth — APPROVED** [review]

File: specification/metrics/supplementary-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+* (T<sub>0</sub>, T<sub>5</sub>]
+  * dimensions: {verb = `GET`, status = `200`}, count: `6`, min: `30 (ms)`, max:
+    `100 (ms)`
+  * dimensions: {verb = `GET`, status = `500`}, count: `3`, min: `1 (ms)`, max:
+    `5 (ms)`
+
+You can see that we are performing Delta->Cumulative conversion, and the SDK
+**has to track what has happened prior to the latest collection/export cycle**,
+in the worst case, the SDK **will have to remember what has happened since the
+ever beginning of the process**.
```
Comment (alanwest) [inline specification/metrics/supplementary-guidelines.md:103]:
```suggestion
beginning of the process**.
```

File: specification/metrics/supplementary-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+  * verb = `GET`, status = `200`, duration = `30 (ms)`
+  * verb = `GET`, status = `200`, duration = `50 (ms)`
+
+Let's imagine we export the metrics as [Histogram](./datamodel.md#histogram),
+and to simplify the story we will only have one histogram bucket `(-Inf, +Inf)`:
+
+If we export the metrics using **Delta Temporality**:
+
+* (T<sub>0</sub>, T<sub>1</sub>]
+  * dimensions: {verb = `GET`, status = `200`}, count: `2`, min: `50 (ms)`, max:
```
Comment (cijothomas) [inline specification/metrics/supplementary-guidelines.md:51]: nit: Histogram also has Sum

**Review by jmacd — APPROVED** [review]
❤️ ❤️ ❤️

File: specification/metrics/supplementary-guidelines.md (hunk 5 above)
Comment (reyang) [reply specification/metrics/supplementary-guidelines.md:51]: Yeah, I think I omitted many things here intentionally so the reader could focus on the key points.
